c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -3296,7 +3298,39 @@ class DeviceCachingAllocator {
       C10_CUDA_CHECK(c10::cuda::SetDevice(stream.device_index()));
 
       EventPool::Event event = create_event_internal(stream.device_index());
-      C10_CUDA_CHECK(cudaEventRecord(*event, stream.stream()));
+      cudaStream_t stream_id = stream.stream();
+
+      if (C10_LIKELY(captures_underway.empty())) {
+        C10_CUDA_CHECK(cudaEventRecordWithFlags(
+            *event, stream_id, cudaEventRecordDefault));
+      } else {
+        // During graph capture, determine if this stream is participating in
+        // any active capture
+        cudaStreamCaptureStatus status{};
+        C10_CUDA_CHECK(cudaStreamGetCaptureInfo(stream_id, &status, nullptr));
+        bool stream_is_capturing = false;
+        for (auto& entry : captures_underway) {
+          if (entry.second(stream_id)) {
+            // Stream is part of an active graph capture - use external
+            // recording to ensure the event is properly included in the graph
+            stream_is_capturing = true;
+            C10_CUDA_CHECK(cudaEventRecordWithFlags(
+                *event, stream_id, cudaEventRecordExternal));
```

Comment:
If you have an external event record here, that would imply that there is no need to rejoin these streams, since there isn't going to be an unjoined fork. Nevertheless, you are adding this stream to `streams_to_rejoin`.

I think your logic around `streams_to_rejoin` is wrong. Bear in mind that it is "correct" (i.e., you won't get an error) to join a stream twice back to its origin stream. I think that is what you are doing, though this isn't a reasonable behavior in this case based on my understanding.

Reply:
Hi, from [this experiment](https://gist.github.com/eee4017/bd9007659b3f5b8633da799cebc464ff), I think the cudaEventRecord do change the bahavior of the stream.